common-rest/src/main/java/org/uniprot/api/rest/service/DefaultSearchQueryOptimiser.java

```diff
@@ -0,0 +1,105 @@
+package org.uniprot.api.rest.service;
+
+import static org.uniprot.core.util.Utils.notNullNotEmpty;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+
+import org.apache.lucene.analysis.core.WhitespaceAnalyzer;
+import org.apache.lucene.index.Term;
+import org.apache.lucene.queryparser.classic.ParseException;
+import org.apache.lucene.queryparser.classic.QueryParser;
+import org.apache.lucene.search.BooleanClause;
+import org.apache.lucene.search.BooleanQuery;
+import org.apache.lucene.search.Query;
+import org.apache.lucene.search.TermQuery;
+import org.uniprot.api.common.exception.ServiceException;
+import org.uniprot.store.config.searchfield.model.SearchFieldItem;
+
+/**
+ * @author lgonzales
+ * @since 10/06/2020
+ */
+public class DefaultSearchQueryOptimiser {
+
+    private final List<SearchFieldItem> optimisedFields;
+
+    public DefaultSearchQueryOptimiser(SearchFieldItem idField) {
+        this(Collections.singletonList(idField));
+    }
+
+    public DefaultSearchQueryOptimiser(List<SearchFieldItem> optimisedFields) {
+        this.optimisedFields = optimisedFields;
+    }
+
+    /**
+     * This method go through requested user query and verify if it can be optimised
+     *
+     * @param requestedQuery requested query
+     * @return Optimised search query string
+     */
+    public String optimiseSearchQuery(String requestedQuery) {
+        String result = requestedQuery;
+        try {
+            QueryParser qp = new QueryParser("", new WhitespaceAnalyzer());
+            qp.setAllowLeadingWildcard(true);
+            Query parsedQuery = qp.parse(requestedQuery);
+            List<String> defaultTerms = getDefaultSearchFieldTerms(parsedQuery);
+            for (String defaultTerm : defaultTerms) {
+                Optional<TermQuery> optimizedQuery = getOptimisedDefaultTermForValue(defaultTerm);
+                if (optimizedQuery.isPresent()) {
+                    if (result.contains("\"" + defaultTerm + "\"")) {
+                        defaultTerm = "\"" + defaultTerm + "\"";
+                    }
+                    result = result.replace(defaultTerm, optimizedQuery.get().toString());
+                }
+            }
+        } catch (ParseException e) {
+            throw new ServiceException(
+                    "Unable to parse query in the optimisation" + requestedQuery, e);
+        }
+        return result;
+    }
+
+    private List<String> getDefaultSearchFieldTerms(Query query) {
```

Comment:
Would it make sense to recursively iterate through `query` (as you already do), then just return the `List<TermQuery>` (call it `X`) whose field names are `""`. Then, `getOptimisedDefaultTermForValue` can mutate the element `e` of `X` if the `e` matches any field in `optimisableFields`'s regexes?

This way we don't do anything special with "" and the query structure isn't being changed -- we're only adding field names where applicable. Maybe this way seems less scary?